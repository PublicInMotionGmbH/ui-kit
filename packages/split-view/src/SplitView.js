import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

import ItemList from './ItemList'
import DetailsView from './DetailsView'
import { moduleName } from './config'

const propTypes = {
  /** Additional class name. */
  className: PropTypes.string,

  /** Data to be displayed inside SplitView component. */
  data: PropTypes.arrayOf(PropTypes.object).isRequired,

  /** Render function of details section. */
  renderDetails: PropTypes.func,

  /** Render function of list section. */
  renderItems: PropTypes.func,

  /** Header element of list section. */
  listHeader: PropTypes.node,

  /** onSelect callback. */
  onSelect: PropTypes.func,

  /** Opened element. */
  openedItem: PropTypes.object
}

/**
 * Component which represents Split View.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {object[]} props.data
 * @property {function} [props.renderDetails]
 * @property {function} [props.renderItems]
 * @property {*} [props.listHeader]
 * @property {function} [props.onSelect]
 * @property {number} [props.openedItem]
 *
 * @property {object} state
 * @property {object} state.openedItem
 *
 * @class
 */
class SplitView extends React.Component {
  state = {
    openedItem: this.props.openedItem || this.props.data[0]
  }

  /**
   * Update index of opened element if it has been changed.
   *
   * @param {object} nextProps
   * @param {object} [nextProps.openedItem]
   */
  componentWillReceiveProps (nextProps) {
    if (nextProps.openedItem !== this.state.openedItem) {
      this.setState({ openedItem: nextProps.openedItem })
    }
  }

  /**
   * Update opnened element if it has been changed by clicking
   * the list element and invoke props.onSelect function.
   *
   * @param {object} item
   */
  onClick = (item, ...args) => {
    const { openedItem } = this.state
    const { onSelect, openedItem: propsOpenedItem } = this.props

    if (onSelect) { onSelect(item, ...args) }
    if (propsOpenedItem == null && openedItem !== item) {
      this.setState({ openedItem: item })
    }
  }
  /**
   * Generates header component if provided.
   */
  generateHeader = () => {
    const { listHeader } = this.props
    const headerCls = buildClassName([moduleName, 'header'], listHeader.props.className)
    return React.cloneElement(listHeader, { className: headerCls })
  }

  render () {
    const { onClick, generateHeader } = this
    const { openedItem } = this.state
    const { className, data, renderDetails, renderItems, listHeader, onSelect, openedItem: propsOpenedItem, ...passedProps } = this.props

    // Class Names
    const wrapperCls = buildClassName(moduleName, className)
    const panelCls = buildClassName([moduleName, 'side-panel'], className)

    return (
      <div className={wrapperCls} {...passedProps}>
        <div className={panelCls}>
          { listHeader && generateHeader() }
          <ItemList
            items={data}
            renderItems={renderItems}
            onClick={onClick}
            openedItem={openedItem}
          />
        </div>
        <DetailsView
          renderDetails={renderDetails}
          openedItem={openedItem}
        />
      </div>
    )
  }
}

SplitView.propTypes = propTypes

export default SplitView
