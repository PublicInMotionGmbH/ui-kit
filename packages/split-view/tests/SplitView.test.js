import React from 'react'
import { shallow } from 'enzyme'

import SplitView from '../src/SplitView'

describe('<SplitView />', () => {
  it('renders children correctly', () => {
    const wrapper = shallow(<SplitView />)
    console.log(wrapper)
    // expect(wrapper).toMatchSnapshot()
  })
})
