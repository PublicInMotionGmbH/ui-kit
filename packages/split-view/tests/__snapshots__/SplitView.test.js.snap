// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsView> when rendered renders children correctly 1`] = `
<div
  className="talixo-split-view"
>
  <div
    className="talixo-split-view__side-panel"
  >
    <ItemList
      items={
        Array [
          Object {
            "name": "Josh Smith I",
            "subtitle": "Subtitle 1",
            "title": "Title 1",
            "value": 0,
          },
          Object {
            "name": "Josh Smith II",
            "subtitle": "Subtitle 2",
            "title": "Title 2",
            "value": 11,
          },
          Object {
            "name": "Josh Smith III",
            "subtitle": "Subtitle 3",
            "title": "Title 3",
            "value": 22,
          },
          Object {
            "name": "Josh Smith IV",
            "subtitle": "Subtitle 4",
            "title": "Title 4",
            "value": 33,
          },
          Object {
            "name": "Josh Smith V",
            "subtitle": "Subtitle 5",
            "title": "Title 5",
            "value": 44,
          },
        ]
      }
      onClick={[Function]}
      openedItem={
        Object {
          "name": "Josh Smith I",
          "subtitle": "Subtitle 1",
          "title": "Title 1",
          "value": 0,
        }
      }
    />
  </div>
  <DetailsView
    openedItem={
      Object {
        "name": "Josh Smith I",
        "subtitle": "Subtitle 1",
        "title": "Title 1",
        "value": 0,
      }
    }
    renderDetails={[Function]}
  />
</div>
`;
